OpenRA.Mods.Common/Traits/BotModules/Squads/States/GroundStates.cs

```diff
@@ -68,57 +84,113 @@ public void Tick(Squad owner)
-				foreach (var unit in owner.Units.Where(a => !ownUnits.Contains(a)))
-					owner.Bot.QueueOrder(new Order("AttackMove", unit, Target.FromCell(owner.World, leader.Location), false));
+				owner.TargetActor = enemyActor;
+				owner.FuzzyStateMachine.ChangeState(owner, new GroundUnitsAttackState(), true);
+				return;
 			}
-			else
+
+			// 2. Make sure the guide unit has not been blocked by the rest of the squad
+			if (attemptsToAdvance <= 0)
 			{
-				var enemies = owner.World.FindActorsInCircle(leader.CenterPosition, WDist.FromCells(owner.SquadManager.Info.AttackScanRadius))
-					.Where(owner.SquadManager.IsPreferredEnemyUnit);
-				var target = enemies.ClosestTo(leader.CenterPosition);
-				if (target != null)
+				if (makeWay > 0)
 				{
-					owner.TargetActor = target;
-					owner.FuzzyStateMachine.ChangeState(owner, new GroundUnitsAttackState(), true);
+					owner.Bot.QueueOrder(new Order("AttackMove", leader, Target.FromCell(owner.World, owner.TargetActor.Location), false));
+
+					foreach (var a in owner.Units)
+					{
+						if (a != leader)
+							owner.Bot.QueueOrder(new Order("Scatter", a, false));
+					}
+
+					makeWay--;
 				}
 				else
-					foreach (var a in owner.Units)
-						owner.Bot.QueueOrder(new Order("AttackMove", a, Target.FromCell(owner.World, owner.TargetActor.Location), false));
+				{
+					// When going through is over, restore the check
+					attemptsToAdvance = MaxAttemptsToAdvance + MakeWayTicks;
+					makeWay = MakeWayTicks;
+				}
+
+				return;
 			}
 
-			if (ShouldFlee(owner))
-				owner.FuzzyStateMachine.ChangeState(owner, new GroundUnitsFleeState(), true);
+			// 3. Check if the squad is stuck due to the map having a very twisted path
+			// or currently bridge and tunnel from TS mod
+			if ((leader.CenterPosition - lastPos).LengthSquared <= 4)
+				attemptsToAdvance--;
+			else
+				attemptsToAdvance = MaxAttemptsToAdvance;
+
+			lastPos = leader.CenterPosition;
+
+			// 4. Since units have different movement speeds, they get separated while approaching the target.
+
+			// Let them regroup into tighter formation towards "leader".
+			//
+			// "groupArea" means the space the squad units will occupy (if 1 per Cell).
+			// leader only stop when scope of "lookAround" is not covered all units;
+			// units in "unitsHurryUp"  will catch up, which keep the team tight while not stuck.
+			//
+			// Imagining "groupArea" takes up a a place shape like square, we need to draw a circle
+			// to cover the the enitire circle.
+			//
+			// PathGuide will check how many squad members are around
+			// to decide if it needs to continue. And units that need hurry up will try
+			// catch up before guide waiting.
+			//
+			// However in practice because of the poor PF, squad tend to PF to a ellipse.
+			// "guideWaitCheck" now has the radius of two times of the circle mentioned before.
+			var occupiedArea = (long)WDist.FromCells(owner.Units.Count).Length * 1024;
```

Comment:
These definitions don't make sense. The variable name suggests it is an area, but the value shows it is a distance, and it is used as if it were a squared distance.  The comments above ("we need to draw a circle to cover the the enitire circle.") doesn't help to clarify the intent.

I can see the general idea here, but I can't suggest specific name/comment changes to make the code more understandable until the logic is fixed to consistently use distance or squared distance or area.

Reply:
change some of the descriptions